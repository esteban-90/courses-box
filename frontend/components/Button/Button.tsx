import type { FC } from 'react'
import styled from '@emotion/styled'
import { makeShadow } from '@/styles/helpers'

type ButtonProps = {
  /** Button action */
  action?: 'primary' | 'secondary' | 'danger' | 'warning'
}

export const Button = styled.button<ButtonProps>`
  all: unset;
  display: flex;
  justify-self: center;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  font-size: 1.6rem;
  width: 15rem;
  height: 4rem;
  border-radius: 1rem;
  transition: all 0.4s ease;

  color: ${({ action, theme }) => {
    const color = action === 'secondary' ? 'regular' : action === 'warning' ? 'warning' : 'button'
    return theme.font[color]
  }};

  background-color: ${({ action = 'primary', theme }) => {
    const backgroundColor = action === 'secondary' ? 'transparent' : theme.components[action]
    return backgroundColor
  }};

  box-shadow: ${({ theme }) => {
    const { shadow1, shadow2 } = theme.components
    const boxShadow = makeShadow(shadow1, shadow2)
    return boxShadow
  }};

  &:active {
    box-shadow: ${({ theme }) => {
      const { shadow1, shadow2 } = theme.components
      const boxShadow = makeShadow(shadow1, shadow2, true)
      return boxShadow
    }};
  }

  &:hover {
    opacity: 0.9;
  }
`

type ActionButtonProps = Omit<ButtonProps, 'action'>

export const PrimaryButton: FC<ActionButtonProps> = (props): JSX.Element => <Button action='primary' {...props} />
export const SecondaryButton: FC<ActionButtonProps> = (props): JSX.Element => <Button action='secondary' {...props} />
export const DangerButton: FC<ActionButtonProps> = (props): JSX.Element => <Button action='danger' {...props} />
export const WarningButton: FC<ActionButtonProps> = (props): JSX.Element => <Button action='warning' {...props} />
