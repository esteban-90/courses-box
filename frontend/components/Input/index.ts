export * from './Input'
export * from './Feedback'
