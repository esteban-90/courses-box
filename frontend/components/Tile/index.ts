export * from './Tile'
export * from './CenteredTile'
