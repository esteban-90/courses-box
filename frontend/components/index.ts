export * from './Button'
export * from './Checkbox'
export * from './Course'
export * from './Icon'
export * from './IconButton'
export * from './Input'
export * from './Layout'
export * from './Link'
export * from './Logo'
export * from './Switch'
export * from './Tile'
