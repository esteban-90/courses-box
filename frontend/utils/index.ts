import qs from 'qs'
import { apiUrl } from '@/config'
import type { CoursePayload, CoursesPayload, ErrorPayload, Course as CourseType } from '@/types'

const getData = async (param: string) => {
  const response = await fetch(`${apiUrl}/courses${param}`, { method: 'GET' })
  const result = await response.json()
  const { status } = response
  const error = status >= 400 ? (result as ErrorPayload) : undefined
  const success = status >= 200 && status < 400 ? (result as CoursesPayload | CoursePayload) : undefined

  return { data: success?.data, error: error?.error?.message ?? '' }
}

type Data<D> = Promise<{ data: D; error: string }>

export const getCourses = () => getData('?populate=*') as Data<CourseType[]>
export const getCourseById = (id: string) => getData(`/${id}?populate=*`) as Data<CourseType>

export const getCoursesByQuery = (q: string) => {
  const query = qs.stringify(
    {
      populate: '*',
      filters: {
        $or: [{ title: { $containsi: q } }, { subtitle: { $containsi: q } }, { description: { $containsi: q } }],
      },
    },
    { encodeValuesOnly: true }
  )

  return getData(`?${query}`) as Data<CourseType[]>
}
