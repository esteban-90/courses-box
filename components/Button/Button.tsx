import { MouseEvent } from 'react'
import styled from '@emotion/styled'
import { boxShadow, transition, getColors, Color } from '@/components/styles'

export type ButtonProps = {
  children: string
  color?: Color
  onClick: (event: MouseEvent<HTMLButtonElement>) => void
}

export const Button = styled.button<ButtonProps>`
  all: unset;
  display: flex;
  justify-self: center;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  font-size: 1.6rem;
  width: 15rem;
  height: 4rem;
  border-radius: 1rem;

  &:hover {
    opacity: 0.9;
  }

  &:active {
    ${({ theme }) => boxShadow(theme.components.shadow1, theme.components.shadow2, true)}
  }

  ${transition()}
  ${({ color, theme }) => getColors(theme, color)}
  ${({ theme }) => boxShadow(theme.components.shadow1, theme.components.shadow2)}
`

Button.defaultProps = { color: 'primary' }
