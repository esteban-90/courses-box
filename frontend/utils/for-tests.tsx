import { FC, ReactElement, ReactNode } from 'react'
import { Provider as StoreProvider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import { render, RenderResult } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'
import { Layout } from '@/components'
import { rootReducer } from '@/store'
import { Themes } from '@/styles/themes'
import { StoreAndRenderOptions } from '@/types'

const customRender = (ui: ReactElement, { preloadedState, ...options }: StoreAndRenderOptions = {}): RenderResult => {
  const store = configureStore({ reducer: rootReducer, preloadedState })

  const Wrapper: FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
    return (
      <StoreProvider store={store}>
        <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
      </StoreProvider>
    )
  }

  return render(ui, { wrapper: Wrapper, ...options })
}

const pageRender = (ui: ReactElement, { preloadedState, ...options }: StoreAndRenderOptions = {}): RenderResult => {
  const store = configureStore({ reducer: rootReducer, preloadedState })

  const Wrapper: FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
    return (
      <StoreProvider store={store}>
        <Layout>{children}</Layout>
      </StoreProvider>
    )
  }

  return render(ui, { wrapper: Wrapper, ...options })
}

export * from '@testing-library/react'
export { customRender as render, pageRender }
