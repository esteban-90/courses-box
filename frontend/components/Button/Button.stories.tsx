import type { ComponentStoryObj, ComponentMeta } from '@storybook/react'
import { screen, userEvent } from '@storybook/testing-library'
import { Button } from '@/components/Button'

export default {
  title: 'Controls/Button',
  component: Button,
  args: {
    children: 'Button',
  },
} as ComponentMeta<typeof Button>

export const PrimaryButton: ComponentStoryObj<typeof Button> = {
  async play() {
    await userEvent.click(screen.getByRole('button'))
  },

  args: {
    action: 'primary',
  },
}

export const SecondaryButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: {
    action: 'secondary',
  },
}

export const WarningButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: {
    action: 'warning',
  },
}

export const DangerButton: ComponentStoryObj<typeof Button> = {
  ...PrimaryButton,
  args: {
    action: 'danger',
  },
}
