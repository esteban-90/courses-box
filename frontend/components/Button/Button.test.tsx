import userEvent from '@testing-library/user-event'
import { Button } from '@/components/Button'
import { render, screen } from '@/tests/utils'

describe('Button test cases:', () => {
  const clickHandler = jest.fn()
  const buttonElement = <Button onClick={clickHandler}>Click here</Button>

  it('should check render', () => {
    const { asFragment } = render(buttonElement)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should check onClick callback', async () => {
    render(buttonElement)
    await userEvent.click(screen.getByRole('button'))
    expect(clickHandler).toHaveBeenCalled()
  })
})
