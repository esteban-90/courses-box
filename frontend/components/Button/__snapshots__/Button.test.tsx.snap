// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test cases: should check render 1`] = `
<DocumentFragment>
  <button
    class="css-tp74ot"
  >
    Click here
  </button>
</DocumentFragment>
`;
